functions/src/index.ts

```diff
@@ -35,85 +36,107 @@ exports.thingspeakToFirestore = functions
-      const thingspeakInfo: PurpleAirResponse = await getThingspeakKeysFromPurpleAir(
-        data['purpleAirId']
-      );
-      const channelAPrimaryData = await axios({
-        url: thingspeakUrl(thingspeakInfo.channelAPrimaryId),
-        params: {
-          api_key: thingspeakInfo.channelAPrimaryKey, // eslint-disable-line camelcase
-          results: 1,
-        },
-      });
-      const channelBPrimaryData = await axios({
-        url: thingspeakUrl(thingspeakInfo.channelBPrimaryId),
-        params: {
-          api_key: thingspeakInfo.channelBPrimaryKey, // eslint-disable-line camelcase
-          results: 1,
-        },
-      });
-      const reading = SensorReading.fromThingspeak(
-        channelAPrimaryData,
-        channelBPrimaryData,
-        thingspeakInfo
-      );
+    for (const sensorDoc of sensorList) {
+      const sensorDocData = sensorDoc.data();
+      const isActive =
+        sensorDocData.isActive === undefined ? true : sensorDocData.isActive;
+      if (isActive && sensorDocData.purpleAirId) {
+        const thingspeakInfo: PurpleAirResponse = await getThingspeakKeysFromPurpleAir(
+          sensorDocData.purpleAirId
+        );
+        const channelAPrimaryData = await axios({
+          url: thingspeakUrl(thingspeakInfo.channelAPrimaryId),
+          params: {
+            api_key: thingspeakInfo.channelAPrimaryKey, // eslint-disable-line camelcase
+            results: 1,
+          },
+        });
+        const channelBPrimaryData = await axios({
+          url: thingspeakUrl(thingspeakInfo.channelBPrimaryId),
+          params: {
+            api_key: thingspeakInfo.channelBPrimaryKey, // eslint-disable-line camelcase
+            results: 1,
+          },
+        });
+        const reading = SensorReading.fromThingspeak(
+          channelAPrimaryData,
+          channelBPrimaryData,
+          thingspeakInfo
+        );
+        const readingTimestamp = Timestamp.fromDate(reading.timestamp);
 
-      const docId = knownSensor.id;
-      const resolvedPath = readingsSubcollection(docId);
-      const readingsRef = firestore.collection(resolvedPath);
+        const readingsCollectionRef = firestore.collection(
+          readingsSubcollection(sensorDoc.id)
+        );
+        let firestoreSafeReading: Pm25BufferElement = defaultPm25BufferElement;
 
-      // If data is not already present in the database add
-      // it to the historical readings
-      let firestoreSafeReading = defaultPm25BufferElement;
-      if (
-        (
-          await readingsRef
-            .where('timestamp', '==', Timestamp.fromDate(reading.timestamp))
+        // If the lastSensorReadingTime field isn't set, query the database to find
+        // the timestamp of the most recent reading. If there are no readings in
+        // Firestore, then lastSensorReading is never changed from null
+        let lastSensorReadingTime: FirebaseFirestore.Timestamp | null =
+          sensorDocData.lastSensorReadingTime ?? null;
```

Comment:
A bit confused by the `?? null` Wouldn't having a null of the left side then coalesce into null on the right? Why do we need to specially do this? Or is this to deal with other nullish things that aren't actually null?

Reply:
The nullish coalescing operator checks for `undefined` or `null` values, and the timestamp can be `undefined` if there is no entry in the document, or `null` if it is defined in the document, but the most recent sensor reading time is unknown (which I use `null` to denote). This line just makes all undefined values `null` so that it's easier to process later, since all of these values are `Timestamp | null` rather than `Timestamp | null | undefined`.